Sources/Flatten.swift

```diff
@@ -733,6 +755,120 @@ private struct LatestState<Value, Error: Swift.Error> {
 	var replacingInnerSignal: Bool = false
 }
 
+extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Error {
+	/// Returns a signal that forwards values from the "first input signal to send an event"
+	/// (winning signal) that is sent on `self`, ignoring values sent from other inner signals.
+	///
+	/// An error sent on `self` or the winning inner signal will be sent on the
+	/// returned signal.
+	///
+	/// The returned signal completes when `self` and the winning inner signal have both completed.
+	fileprivate func first() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { observer in
+			let composite = CompositeDisposable()
+			let relayDisposable = CompositeDisposable()
+
+			composite += relayDisposable
+			composite += self.observeFirst(observer, relayDisposable)
+
+			return composite
+		}
+	}
+
+	fileprivate func observeFirst(_ observer: Observer<Value.Value, Error>, _ relayDisposable: CompositeDisposable) -> Disposable? {
+		let state = Atomic(FirstState<Value.Value, Error>())
+
+		return self.observe { event in
+			switch event {
+			case let .value(innerProducer):
+				// Ignore consecutive `innerProducer`s if first `innerSignal` already sent an event.
+				guard !relayDisposable.isDisposed else {
+					return
+				}
+
+				innerProducer.startWithSignal { innerSignal, innerDisposable in
+
+					let disposableHandle = relayDisposable.add(innerDisposable)
+
+					innerSignal.observe { [unowned innerSignal] event in
+
+						let isFirstSignal: Bool = state.modify {
+							if $0.active == nil {
+								$0.active = innerSignal
+							}
+							return $0.active === innerSignal
+						}
+
+						// Ignore non-winning signals.
+						guard isFirstSignal else { return }
+
+						// Dispose all running innerSignals except winning one.
+						disposableHandle.remove()
+						relayDisposable.dispose()
+
+						switch event {
+						case .completed:
+							let shouldComplete: Bool = state.modify {
+								$0.active = nil
+								return $0.outerSignalComplete
+							}
+
+							if shouldComplete {
+								observer.sendCompleted()
+							}
+
+						case .value, .failed, .interrupted:
+							observer.action(event)
+						}
+					}
+				}
+
+			case let .failed(error):
+				observer.send(error: error)
+
+			case .completed:
+				let shouldComplete: Bool = state.modify {
+					$0.outerSignalComplete = true
+					return $0.active == nil
```

Comment:
I haven't written the test yet, but I think the result should be the same as:
	
`it("should complete when the outer signal completes before sending any signals")`.
https://github.com/ReactiveCocoa/ReactiveSwift/pull/233/files#diff-9cd786f9e52e786771d96cf8c42d3befR1481

New test will be: `it("should complete when the outer signal completes before inner signals send any event")`.